### PR TITLE
⚠️ CONFLICT! Lineage pull request for: skeleton

### PR DESCRIPTION
**Pull request by cisagovbot**
# Lineage Pull Request: CONFLICT #

Achtung!!!

[Lineage] has created this pull request to incorporate new changes found in an upstream repository:

Upstream repository: [`https://github.com/cisagov/skeleton-python-library.git`](https://github.com/cisagov/skeleton-python-library.git)
Remote branch: `HEAD`

Check the changes in this pull request to ensure they won't cause issues with your project.

The `lineage/skeleton` branch has **one or more unresolved merge conflicts** that you must resolve before merging this pull request!

## How to resolve the conflicts ##

1. Take ownership of this pull request by removing any other assignees.

1. Clone the repository locally, and reapply the merge:

    ```console
    git clone git@github.com:cisagov/cyhy-db.git cyhy-db
    cd cyhy-db
    git remote add skeleton https://github.com/cisagov/skeleton-python-library.git
    git remote set-url --push skeleton no_push
    git switch develop
    git switch --create lineage/skeleton --track origin/develop
    git pull skeleton HEAD
    git status
    ```

1. Review the changes displayed by the `status` command.  Fix any conflicts and possibly incorrect auto-merges.

1. After resolving each of the conflicts, `add` your changes to the branch, `commit`, and `push` your changes:

    ```console
    git add src/example/example.py 
    git commit
    git push --force --set-upstream origin lineage/skeleton
    ```

    Note that you may *append* to the default merge commit message that git creates for you, but *please do not delete the existing content*.  It provides useful information about the merge that is being performed.

1. Wait for all the automated tests to pass.

1. Confirm each item in the "Pre-approval checklist" below.

1. Remove any of the checklist items that do not apply.

1. Ensure every remaining checkbox has been checked.

1. Mark this draft pull request "Ready for review".

## ✅ Pre-approval checklist ##

- [x] ✌️ The conflicts in this pull request have been resolved.
- [x] All relevant type-of-change labels have been added.
- [x] All relevant repo and/or project documentation has been updated to reflect the changes in this PR.
- [x] Tests have been added and/or modified to cover the changes in this PR.
- [x] All new and existing tests pass.
- [x] Bump major, minor, patch, pre-release, and/or build versions [as appropriate](https://semver.org/#semantic-versioning-specification-semver) via the `bump_version` script *if* this repository is versioned *and* the changes in this PR [warrant a version bump](https://semver.org/#what-should-i-do-if-i-update-my-own-dependencies-without-changing-the-public-api).
- [x] Create a pre-release (necessary if and only if the pre-release version was bumped).

## ✅ Pre-merge checklist ##

- [x] Finalize version.

## ✅ Post-merge checklist ##

- [x] Create a release (necessary if and only if the version was bumped).

---

> [!NOTE]
> You are seeing this because one of this repository's maintainers has configured [Lineage] to open pull requests.

For more information:

🛠 [Lineage] configurations for this project are stored in `.github/lineage.yml`

📚 [Read more about Lineage][Lineage]

[//]: # (lineage:metadata:{&quot;slayed&quot;:true})
[Lineage]: https://github.com/cisagov/action-lineage/ "Lineage GitHub Action"

### CODE REVIEW

**Comment by jsf9k** [comment]
@felddy and @dav3r - Can you help me determine if the remaining `flake8` failures in the lint job are necessary because of `beanie` or whatever?  If they are then I'll add `noqa` comments, but if they aren't then I'd like to clean them up.

**Comment by dav3r** [comment]
> @felddy and @dav3r - Can you help me determine if the remaining `flake8` failures in the lint job are necessary because of `beanie` or whatever? If they are then I'll add `noqa` comments, but if they aren't then I'd like to clean them up.

For the `N805` errors, the AI overlords suggest fixing it by putting this in our `pyproject.toml`:

```
[tool.flake8]
classmethod-decorators = [
    "classmethod",
    "model_validator",
    "validator",
    "field_validator",
    "root_validator"
]
```

Here's the relevant info I found:

> In Pydantic v2 (which Beanie uses), `@model_validator` methods function like class methods and typically take `cls` as the first argument. However, flake8-pep8-naming assumes any method not explicitly decorated with `@classmethod` is an instance method and therefore expects the first argument to be `self`.

> ~Flake8 does not natively support `pyproject.toml`. To use it, you need to install a helper plugin like [Flake8-pyproject](https://pypi.org/project/Flake8-pyproject/).~
> 
> ~Once the plugin is installed,~ add the following section to your pyproject.toml to tell the pep8-naming plugin that `@model_validator` is a class-method-style decorator:
> 
> ```
> [tool.flake8]
> classmethod-decorators = [
>     "classmethod",
>     "model_validator",
>     "validator",
>     "field_validator",
>     "root_validator"
> ]
> ```
> 
> Summary of Steps
> - ~Install the plugin: Run `pip install flake8-pyproject`.~
> - Add the config into your pyproject.toml.
> - Run Flake8: Use your standard flake8 command; the plugin will automatically detect and load these settings.
>  
> If you prefer not to add a new dependency, you can simply add @classmethod above your @model_validator in the code, as Pydantic supports stacking them.

DR edit: I modified above to strike thru the info about adding `flake8-pyproject` as a dependency; @jsf9k verified that it was not necessary in our particular case.  YMMV.

**Comment by felddy** [comment]
To address: `N805 first argument of a method should be named 'self'`:

I think adding the `classmethod-decorators` jazz to `pyproject.toml` is the correct move here.  We're going to see this more in the future.   

I'm looking into the other two findings: `B042`, and `N801`

**Comment by felddy** [comment]
The ```B042 Exception class with `__init__` should pass all args to `super().__init__()` to work in edge cases of `pickle` and `copy.copy()`. It should also not take any kwargs.``` are legit. 

I can push a fix after I test locally.

**Comment by felddy** [comment]
@jsf9k fixes pushed up.  Looking good!

File: src/cyhy_db/models/exceptions.py [hunk 1]
```diff
@@ -5,39 +5,45 @@ class PortScanNotFoundError(Exception):
...
         self.port_scan_id = port_scan_id
         self.port_scan_time = port_scan_time
-        super().__init__(message, *args)
+        super().__init__(ticket_id, port_scan_id, port_scan_time, *args)
+
+    def __str__(self):
+        """Return a human-readable description of the exception."""
+        return "Ticket {}: referenced PortScanDoc {} at time {} not found".format(
+            self.ticket_id, self.port_scan_id, self.port_scan_time
+        )
```
Comment (jsf9k) [inline src/cyhy_db/models/exceptions.py:26]: Isn't this code Python 3?  Why not use an f-string?

File: src/cyhy_db/models/exceptions.py [hunk 2]
```diff
@@ -5,39 +5,45 @@ class PortScanNotFoundError(Exception):
...
         self.vuln_scan_id = vuln_scan_id
         self.vuln_scan_time = vuln_scan_time
-        super().__init__(message, *args)
+        super().__init__(ticket_id, vuln_scan_id, vuln_scan_time, *args)
+
+    def __str__(self):
+        """Return a human-readable description of the exception."""
+        return "Ticket {}: referenced VulnScanDoc {} at time {} not found".format(
+            self.ticket_id, self.vuln_scan_id, self.vuln_scan_time
+        )
```
Comment (jsf9k) [inline src/cyhy_db/models/exceptions.py:51]: Same question applies here.

**Review by jsf9k — CHANGES_REQUESTED** [review]
@felddy, I am confused about the use of old Python 2 string formatting.

File: src/cyhy_db/models/exceptions.py (hunk 1 above)
Comment (felddy) [reply src/cyhy_db/models/exceptions.py:26]: Brain toot from working with Py2 all week.  Fixed in https://github.com/cisagov/cyhy-db/pull/21/commits/1847669d8c82524a1693c11b681970bf0ade512f

File: src/cyhy_db/models/exceptions.py (hunk 2 above)
Comment (felddy) [reply src/cyhy_db/models/exceptions.py:51]: Same response here.

**Review by jsf9k — APPROVED** [review]

**Comment by jsf9k** [comment]
@cisagov/vm-dev - Do you think we need to bump the patch version here since we made so many (minor) code changes?  I was thinking not but wanted to get consensus from the group.

File: src/cyhy_db/models/scan_doc.py [hunk 3]
```diff
@@ -48,7 +48,7 @@ class Settings:
 
         name = "PyTest_ScanDocs"
 
-    class Abstract_Settings:
+    class AbstractSettings:
         """Beanie settings to be inherited by subclasses."""
```
[the remark this answers is not in the document]
Comment (jsf9k) [reply src/cyhy_db/models/scan_doc.py:52]: Please see commit 94bd061.

File: src/cyhy_db/models/exceptions.py [hunk 4]
```diff
@@ -1,43 +1,51 @@
 """The exceptions used in CyHy."""
 
 
-class PortScanNotFoundException(Exception):
+class PortScanNotFoundError(Exception):
```
Comment (dav3r) [inline src/cyhy_db/models/exceptions.py:4]: Not quite sure where we should track this, but the original name (`PortScanNotFoundException`)  is [used in cyhy-reports](https://github.com/cisagov/cyhy-reports/blob/7a83cb7bea8dcc03c496200c394d4d5ecb7b8642/cyhy_report/cyhy_notification/generate_notification.py#L327), so we will want to make a note to change that whenever we switch over to using cyhy-db.

It's also used in cyhy-core [here](https://github.com/cisagov/cyhy-core/blob/2eb4826fec4468775ceebad705e6b5f1f660ac9d/cyhy/db/database.py#L587).

File: src/cyhy_db/models/exceptions.py [hunk 5]
```diff
@@ -1,43 +1,51 @@
...
+    def __str__(self):
+        """Return a human-readable description of the exception."""
+        return (
+            f"Ticket {self.ticket_id}: referenced PortScanDoc {self.port_scan_id} "
+            f"at time {self.port_scan_time} not found"
+        )
 
 
-class VulnScanNotFoundException(Exception):
+class VulnScanNotFoundError(Exception):
```
Comment (dav3r) [inline src/cyhy_db/models/exceptions.py:29]: Not quite sure where we should track this, but the original name (`VulnScanNotFoundException`)  is [used in cyhy-reports](https://github.com/cisagov/cyhy-reports/blob/7a83cb7bea8dcc03c496200c394d4d5ecb7b8642/cyhy_report/cyhy_notification/generate_notification.py#L308), so we will want to make a note to change that whenever we switch over to using cyhy-db.

It's also used in cyhy-core [here](https://github.com/cisagov/cyhy-core/blob/2eb4826fec4468775ceebad705e6b5f1f660ac9d/cyhy/db/database.py#L607).

File: src/cyhy_db/models/exceptions.py (hunk 4 above)
Comment (jsf9k) [reply src/cyhy_db/models/exceptions.py:4]: I think we should create issues in those repos.

**Comment by felddy** [comment]
> @cisagov/vm-dev - Do you think we need to bump the patch version here since we made so many (minor) code changes? I was thinking not but wanted to get consensus from the group.

@jsf9k I think a patch bump will suffice.  

I would have said major, if any of the newer repos were referencing the renamed (🙄 ) exceptions.  [They are not](https://github.com/search?q=org%3Acisagov++VulnScanNotFoundException+PortScanNotFoundException&type=code).

File: src/cyhy_db/models/exceptions.py (hunk 4 above)
Comment (jsf9k) [reply src/cyhy_db/models/exceptions.py:4]: @felddy pointed out that the code will break when we start using this project, so we need not even create an issue.

File: src/cyhy_db/models/exceptions.py (hunk 5 above)
Comment (jsf9k) [reply src/cyhy_db/models/exceptions.py:29]: @felddy pointed out that the code will break when we start using this project, so we need not even create an issue.

Comment (dav3r) [reply src/cyhy_db/models/exceptions.py:29]: I suppose that will work.  Anyway, I'm sure that we will remember this. 🤣

**Review by dav3r — APPROVED** [review]
Looks great, thanks for tackling this one! ⚡
